### PR TITLE
fix: a11y aria-labelledby filter offcanvas

### LINKED ISSUE #7051: A11Y: Filter function not sufficiently labelled
*PHP Version:* 8.2   | *Shopware Version:* 6.6.9.0  | *Affected area / extension:* Platform(Default) 

 ----

*_Actual behaviour:_*

The filter function for mobile is not sufficiently labelled.
Labelling conveys that the elements belong together and screen reader users can jump to this section more quickly.

Image

*_Expected behaviour:_*

There are several ways to mark functions as belonging together.
The simplest is to mark the filter function as navigation and set the text Restrict selection as the navigation label.

```html
<nav arialabedby="filter-heading"> 
 <div id="filter-heading">Filter</div>  
 <div class="filter-panel">   
  <button aria-expanded="false">Hersteller</button>   
   ... 
 </div>
</nav>
```

*_How to reproduce:_*

Enable `ACCESSIBILITY_TWEAKS` Feature-Flag
- Create a sample product and add it to the category listing page. Then a filter is visible.

### PR DESCRIPTION
**Pull request by BrocksiNet**
### 1. Why is this change necessary?
An `aria-labelledby` is currently missing for the filter panel in OffCanvas (mobile).

### 2. What does this change do, exactly?
- Adds the `aria-labelledby` to the Twig template
- Updates the js plugin to add wrapper div only for the offcanvas filter
- Removes some not needed SCSS conditions

### 3. Describe each step to reproduce the issue or behaviour.
- Check the DOM structure for desktop view without OffCanvas
- Check the DOM structure for mobile view with OffCanvas opened

### 4. Please link to the relevant issues (if any).

- closes #7051
- Jira issue: https://shopware.atlassian.net/browse/NEXT-40818

### 5. Checklist

- [X] I have written tests and verified that they fail without my change
- [X] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfill them.

### CODE REVIEW

**Review by Phil23 — APPROVED** [review]

File: src/Storefront/Resources/views/storefront/element/cms-element-sidebar-filter.html.twig [hunk 1]
```diff
@@ -28,8 +28,9 @@
             {% block element_sidebar_filter_panel %}
                 <div
                     id="filter-panel-wrapper"
-                    class="filter-panel-wrapper"
+                    class="d-none d-lg-block"
```
Comment (tobiasberge) [inline src/Storefront/Resources/views/storefront/element/cms-element-sidebar-filter.html.twig:31]: If we use utility classes we should keep the selector classes in place so they can still be customized. 

In this case, we could probably keep the CSS of `filter-panel-wrapper` if we can avoid the JS of `classesToRemoveFromWrapperInOffCanvas`?

**Review by tobiasberge — COMMENTED** [review]
@BrocksiNet Maybe we can find a more general solution here in the `OffCanvas` class. Then it would also work in other OffCanvas that currently don't have a labelledby attribute.

I was thinking of this: 

* We determine an identifier that the `OffCanvas` should automatically look for during `OffCanvas.open()`. 
* The identifier is added in twig.
* The `OffCanvas` class looks inside the OffCanvas content HTML. And when this is identifier is found, we automatically set `aria-labelledby` to this identifier. 
* To avoid duplicate ids, we use a data-attribute and only set the real ID with JS during opening. 

Set data-id to filter panel: `Resources/views/storefront/component/listing/filter-panel.html.twig`
```html
<h2 data-id="off-canvas-headline" class="filter-panel-offcanvas-only filter-panel-offcanvas-title">
  {{ "listing.filterTitleText"|trans }}
</h2>
```

Set data-id to other off-canvas content like "cart": `Resources/views/storefront/component/checkout/offcanvas-cart.html.twig`

```html
<h4 data-id="off-canvas-headline">
  {{- 'checkout.cartHeader'|trans|sw_sanitize -}}
</h4>
```

Now we add a new general method to `Resources/app/storefront/src/plugin/offcanvas/offcanvas.plugin.js`

```diff
    setContent(content, delay) {
        const offCanvas = this.getOffCanvas();

        if (!offCanvas[0]) {
            return;
        }

        offCanvas[0].innerHTML = content;

        // register events again
        this._registerEvents(delay);
+        this._setAriaAttrs();
    }

```

```js
_setAriaAttrs() {
    const offCanvas = this.getOffCanvas()[0];

    const headlineId = 'off-canvas-headline';
    
    // We are only looking inside the offCanvas HTML
    const headline = offCanvas.querySelector(`[data-id="${headlineId}"]`);

    // We search if we can find the data identifier
    if (headline) {
        // If we find it, we set the real id (this only happens in the OffCanvas HTML, thus no duplicate ID)
        headline.setAttribute('id', headlineId);

        // If we find it, we also set the aria-labelledby to the ID.
        offCanvas.setAttribute('aria-labelledby', headlineId);
    }
    // If we don't find the data-id nothing happens like before.
}
```

```
// No changes needed in offcanvas-filter.plugin.js
```

This should work for different OffCanvas, and we don't have duplicate IDs or wrapper divs. I haven't tested yet with screen reader if there are any timing problems because the real id is not in the initial DOM. Maybe you could check this. But since this happens before the OffCanvas opens it should be ok. 

image

image

EDIT: Haha 😸 When you open and close the filter panel multiple times you can have a duplicate id. Because they move the HTML from the OffCanvas back to the original place on close. But should be not a big deal we can remove the id again inside `OffCanvas.close()` method. And this should not matter anyway for ajax modals like cart because the content is always loaded new via ajax.

**Comment by BrocksiNet** [comment]
I close this PR in flavor of https://github.com/shopware/shopware/pull/7279
